CachedResponse.history is not fully deserialized on python 3.7 and 3.8

Found during #835; related to https://github.com/python-attrs/cattrs/issues/206

Example error:
```python
requests_cache/session.py:103: in get
    return self.request('GET', url, params=params, **kwargs)
requests_cache/session.py:159: in request
    return super().request(method, url, *args, headers=headers, **kwargs)  # type: ignore
.venv/lib/python3.7/site-packages/requests/sessions.py:589: in request
    resp = self.send(prep, **send_kwargs)
requests_cache/session.py:195: in send
    actions.update_from_cached_response(cached_response, self.cache.create_key, **kwargs)
requests_cache/policy/actions.py:194: in update_from_cached_response
    elif create_key and not self._validate_vary(cached_response, create_key, **key_kwargs):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = CacheActions(expire_after=-1)
cached_response = CachedResponse(_content=b'mock redirect response with Vary header (2/2)', created_at='2023-06-05 17:56:56.815238', ela.../requests-cache.com/vary-redirect-target'), status_code=200, url='http+mock://requests-cache.com/vary-redirect-target')
create_key = <bound method BaseCache.create_key of <SQLiteCache(name=/tmp/pytest-of-runner/pytest-0/popen-gw1/test_match_headers__vary_with_4/24501901-26eb-4b6b-a6d0-6b12df83c48b.db)>>
key_kwargs = {'allow_redirects': True, 'cert': None, 'match_headers': ['Accept-Language'], 'proxies': OrderedDict(), ...}
vary = 'Accept-Language'

    def _validate_vary(
        self, cached_response: 'CachedResponse', create_key: KeyCallback, **key_kwargs
    ) -> bool:
        """If the cached response contains Vary, check that the specified request headers match"""
        vary = cached_response.headers.get('Vary')
        if not vary:
            return True
        elif vary == '*':
            return False
    
        # Generate a secondary cache key based on Vary for both the cached request and new request.
        # If there are redirects, compare the new request against the last request in the chain.
        key_kwargs['match_headers'] = [k.strip() for k in vary.split(',')]
        vary_request = (
            cached_response.history[-1].request
>           if cached_response.history
            else cached_response.request
        )
E       AttributeError: 'dict' object has no attribute 'request'

requests_cache/policy/actions.py:308: AttributeError
```

`CachedResponse.history` should contain `CachedResponse` when there is redirect history available. Instead, it contains dicts, meaning the structure hook is not working.

--- a/HISTORY.md
+++ b/HISTORY.md
@@ -13,6 +13,7 @@
 * Fix loading cached JSON content when `decode_content=True` and the root element is a list
 * Fix `BaseCache.recreate_keys()` to normalize response bodies with `b'None'`
 * Fix potential `OperationalError: database is locked` during bulk delete operations
+* Fix `CachedResponse.history` not being fully deserialized on python<=3.8
 * Fix request matching with `Vary` and redirects
 * Add compatibility with urllib3 2.0
 

--- a/HISTORY.md
+++ b/HISTORY.md
@@ -13,6 +13,7 @@
 * Fix loading cached JSON content when `decode_content=True` and the root element is a list
 * Fix `BaseCache.recreate_keys()` to normalize response bodies with `b'None'`
 * Fix potential `OperationalError: database is locked` during bulk delete operations
+* Fix request matching with `Vary` and redirects
 * Add compatibility with urllib3 2.0
 
 ## 1.0.1 (2023-03-24)

--- a/requests_cache/policy/actions.py
+++ b/requests_cache/policy/actions.py
@@ -296,9 +296,15 @@ class CacheActions(RichMixin):
         elif vary == '*':
             return False
 
-        # Generate a secondary cache key based on Vary for both the cached request and new request
+        # Generate a secondary cache key based on Vary for both the cached request and new request.
+        # If there are redirects, compare the new request against the last request in the chain.
         key_kwargs['match_headers'] = [k.strip() for k in vary.split(',')]
-        vary_cache_key = create_key(cached_response.request, **key_kwargs)
+        vary_request = (
+            cached_response.history[-1].request
+            if cached_response.history
+            else cached_response.request
+        )
+        vary_cache_key = create_key(vary_request, **key_kwargs)
         headers_match = create_key(self._request, **key_kwargs) == vary_cache_key
         if not headers_match:
             _log_vary_diff(

--- a/requests_cache/serializers/cattrs.py
+++ b/requests_cache/serializers/cattrs.py
@@ -117,6 +117,12 @@ def init_converter(
         DecodedContent, lambda obj, cls: json.dumps(obj) if isinstance(obj, (dict, list)) else obj
     )
 
+    def structure_fwd_ref(obj, cls):
+        # python<=3.8: ForwardRef may not have been evaluated yet
+        if not cls.__forward_evaluated__:  # pragma: no cover
+            cls._evaluate(globals(), locals())
+        return converter.structure(obj, cls.__forward_value__)
+
     # Resolve forward references (required for CachedResponse.history)
     converter.register_unstructure_hook_func(
         lambda cls: cls.__class__ is ForwardRef,
@@ -124,7 +130,7 @@ def init_converter(
     )
     converter.register_structure_hook_func(
         lambda cls: cls.__class__ is ForwardRef,
-        lambda obj, cls: converter.structure(obj, cls.__forward_value__),
+        structure_fwd_ref,
     )
 
     return converter

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -86,6 +86,8 @@ MOCKED_URL_JSON_LIST = 'http+mock://requests-cache.com/json_list'
 MOCKED_URL_REDIRECT = 'http+mock://requests-cache.com/redirect'
 MOCKED_URL_REDIRECT_TARGET = 'http+mock://requests-cache.com/redirect_target'
 MOCKED_URL_VARY = 'http+mock://requests-cache.com/vary'
+MOCKED_URL_VARY_REDIRECT = 'http+mock://requests-cache.com/vary-redirect'
+MOCKED_URL_VARY_REDIRECT_TARGET = 'http+mock://requests-cache.com/vary-redirect-target'
 MOCKED_URL_404 = 'http+mock://requests-cache.com/nonexistent'
 MOCKED_URL_500 = 'http+mock://requests-cache.com/answer?q=this-statement-is-false'
 MOCKED_URL_200_404 = 'http+mock://requests-cache.com/200-404'
@@ -218,14 +220,14 @@ def get_mock_adapter() -> Adapter:
         ANY_METHOD,
         MOCKED_URL_REDIRECT,
         headers={'Content-Type': 'text/plain', 'Location': MOCKED_URL_REDIRECT_TARGET},
-        text='mock redirect response',
+        text='mock redirect response (1/2)',
         status_code=302,
     )
     adapter.register_uri(
         ANY_METHOD,
         MOCKED_URL_REDIRECT_TARGET,
         headers={'Content-Type': 'text/plain'},
-        text='mock redirected response',
+        text='mock redirect response (2/2)',
         status_code=200,
     )
     adapter.register_uri(
@@ -233,6 +235,24 @@ def get_mock_adapter() -> Adapter:
         MOCKED_URL_VARY,
         headers={'Content-Type': 'text/plain', 'Vary': 'Accept'},
         text='mock response with Vary header',
+        status_code=200,
+    )
+    adapter.register_uri(
+        ANY_METHOD,
+        MOCKED_URL_VARY_REDIRECT,
+        headers={
+            'Content-Type': 'text/plain',
+            'Vary': 'Accept',
+            'Location': MOCKED_URL_VARY_REDIRECT_TARGET,
+        },
+        text='mock redirect response with Vary header (1/2)',
+        status_code=302,
+    )
+    adapter.register_uri(
+        ANY_METHOD,
+        MOCKED_URL_VARY_REDIRECT_TARGET,
+        headers={'Content-Type': 'text/plain', 'Vary': 'Accept-Language'},
+        text='mock redirect response with Vary header (2/2)',
         status_code=200,
     )
     adapter.register_uri(ANY_METHOD, MOCKED_URL_404, status_code=404)

--- a/tests/unit/test_session.py
+++ b/tests/unit/test_session.py
@@ -31,6 +31,8 @@ from tests.conftest import (
     MOCKED_URL_REDIRECT,
     MOCKED_URL_REDIRECT_TARGET,
     MOCKED_URL_VARY,
+    MOCKED_URL_VARY_REDIRECT,
+    MOCKED_URL_VARY_REDIRECT_TARGET,
     START_DT,
     ignore_deprecation,
     patch_normalize_url,
@@ -330,15 +332,39 @@ def test_match_headers__vary(mock_session):
     """Vary should be used to validate headers, if available.
     It should also override `match_headers` for the secondary cache key, if both are provided.
     """
-    # mock_session.settings.match_headers = ['Accept-Encoding']
     headers_1 = {'Accept': 'application/json', 'User-Agent': 'qutebrowser'}
     headers_2 = {'Accept': 'application/json', 'User-Agent': 'Firefox'}
     headers_3 = {'Accept': 'text/plain', 'User-Agent': 'qutebrowser'}
+    mock_session.get(MOCKED_URL_VARY, headers=headers_1)
 
-    assert mock_session.get(MOCKED_URL_VARY, headers=headers_1).from_cache is False
     assert mock_session.get(MOCKED_URL_VARY, headers=headers_1).from_cache is True
     assert mock_session.get(MOCKED_URL_VARY, headers=headers_2).from_cache is True
     assert mock_session.get(MOCKED_URL_VARY, headers=headers_3).from_cache is False
+
+
+@pytest.mark.parametrize(
+    'headers, expected_from_cache',
+    [
+        ({'Accept': 'text/plain', 'Accept-Language': 'en-US'}, True),
+        ({'Accept-Language': 'en-US'}, True),
+        ({'Accept-Language': 'en-GB'}, False),
+        ({'Accept': 'application/json'}, False),
+        ({}, False),
+    ],
+)
+def test_match_headers__vary_with_redirects(headers, expected_from_cache, mock_session):
+    """With Vary + redirects, Vary should match against the last request in the redirect chain.
+    * 1st response has `Vary: Accept` (shouldn't be matched)
+    * 2nd response has `Vary: Accept-Language` (should be matched)
+    """
+    r = mock_session.get(
+        MOCKED_URL_VARY_REDIRECT,
+        headers={'Accept': 'text/plain', 'Accept-Language': 'en-US'},
+    )
+    assert r.url == MOCKED_URL_VARY_REDIRECT_TARGET
+
+    r2 = mock_session.get(MOCKED_URL_VARY_REDIRECT, headers=headers)
+    assert r2.from_cache is expected_from_cache
 
 
 def test_include_get_headers():

--- a/tests/unit/test_session.py
+++ b/tests/unit/test_session.py
@@ -19,6 +19,7 @@ from requests_cache import ALL_METHODS, CachedSession
 from requests_cache._utils import get_placeholder_class
 from requests_cache.backends import BACKEND_CLASSES, BaseCache
 from requests_cache.backends.base import DESERIALIZE_ERRORS
+from requests_cache.models import CachedResponse
 from requests_cache.policy.expiration import DO_NOT_CACHE, EXPIRE_IMMEDIATELY, NEVER_EXPIRE
 from tests.conftest import (
     MOCKED_URL,
@@ -200,11 +201,19 @@ def test_verify(mock_session):
 
 
 def test_response_history(mock_session):
+    """When a request results in redirects:
+    * An alias for the redirect should be added to 'redirects' table/collection
+    * Both the original and target URLs should be retrievable from the cache
+    * CachedResponse.history should be populated with CachedResponse objects
+    """
     mock_session.get(MOCKED_URL_REDIRECT)
     r = mock_session.get(MOCKED_URL_REDIRECT_TARGET)
 
     assert r.from_cache is True
     assert len(mock_session.cache.redirects) == 1
+
+    r2 = mock_session.get(MOCKED_URL_REDIRECT)
+    assert isinstance(r2.history[0], CachedResponse)
 
 
 # Request matching

